### PR TITLE
typeck/type_of: let wfcheck handle generics in opaque types' substs.

### PR DESCRIPTION
**Pull request by eddyb**
I was working on #70164, and `type_of`'s handling of opaque types seemed to be, by far, the trickiest use of `Ty::walk`, but I believe it wasn't doing anything (see https://github.com/rust-lang/rust/pull/57896#discussion_r396064431 - I suspect, based on glancing at the PR discussion, that an early attempt was kept in, despite becoming just an overcomplicated way to do exactly the same as the previous simple type equality check).

I would've loved to remove `ResolvedOpaqueTy` (keep the `Ty` and lose the `Substs`), but it looks like the MIR borrowck part of the process needs that now, so it would've been added anyway since #57896, even if that PR hadn't happened.

<hr/>

In the process, I've moved the remaining substitution validation to `wfcheck`, which was already handling lifetimes, and kept only `delay_span_bug`s in `type_of`, as an insurance policy.

I've added tests for lifetime and const cases, they seem to be checked correctly now.
(and more uniform than they were in https://github.com/rust-lang/rust/issues/63063#issuecomment-602162804)

However, the quality of the errors is maybe a bit worse, and they don't trigger when there are other errors (not sure if this is due to compilation stop points or something more specific to one opaque type).

r? @nikomatsakis cc @matthewjasper @oli-obk @Aaron1011

### CODE REVIEW

File: src/librustc_typeck/collect/type_of.rs
Comment (nikomatsakis) [inline src/librustc_typeck/collect/type_of.rs]: Pre-existing, but this code is in dire need of some comments.

To start, can we add a comment explaining what is happening here -- I think it is this:

---

Examine the type parameters to the opaque type used in the *defining use* and make sure that (a) they are only generic parameters from the enclosing scope and (b) each generic parameter is used at most once. For example, given this function:

```rust
type Foo<A, B> = impl Debug;

fn f1<T, U>() -> Foo<T, U>
```

we would be iterating over the substs `T, U` and checking that they are distinct. We would error for things like `-> Foo<T, T>` or `-> Foo<u32, U>`.

**Review by nikomatsakis — COMMENTED** [review]
started reading but ran out of time. =) will return. initial comment.

File: src/librustc_typeck/collect/type_of.rs
Comment (eddyb) [reply src/librustc_typeck/collect/type_of.rs]: `wfcheck` is the primary checker and has a nice comment.
This is just a tombstone insurance policy. But I can try to document it if you want.

Comment (nikomatsakis) [reply src/librustc_typeck/collect/type_of.rs:401]: I'd still like a comment, even if only points over to the "primary comment"

File: src/test/ui/type-alias-impl-trait/generic_duplicate_param_use2.stderr
Comment (nikomatsakis) [inline src/test/ui/type-alias-impl-trait/generic_duplicate_param_use2.stderr:1]: This *does* seem like more confusing output than the old code...

File: src/test/ui/type-alias-impl-trait/generic_nondefining_use.stderr
Comment (nikomatsakis) [inline src/test/ui/type-alias-impl-trait/generic_nondefining_use.stderr:13]: This error message seems pretty confusing to me. But I guess the old one was pretty bad too so I don't think it needs to block this PR...

File: src/librustc_typeck/collect/type_of.rs
Comment (nikomatsakis) [inline src/librustc_typeck/collect/type_of.rs:408]: I guess the other types just didn't happen in practice? Else it'd be nice to have a test.

**Review by nikomatsakis — APPROVED** [review]
r=me with nits addressed (I think just a comment)

File: src/librustc_typeck/collect/type_of.rs
Comment (nikomatsakis) [reply src/librustc_typeck/collect/type_of.rs:401]: In fact, I think pointing out that this is "just" an insurance policy would be quite useful =)

File: src/librustc_typeck/collect/type_of.rs
Comment (eddyb) [reply src/librustc_typeck/collect/type_of.rs:408]: These are "exported" by the MIR borrowck (cc @matthewjasper) AFAICT, and that results in late-bound lifetimes (in the `fn` signature, I presume) being `ReFree`.

This list is minimal, as in I can't remove either `ReEarlyBound` nor `ReFree` without tests failing, and adding more would cause less errors (i.e. allowing buggy situations), not more.

File: src/test/ui/type-alias-impl-trait/generic_duplicate_param_use2.stderr
Comment (eddyb) [reply src/test/ui/type-alias-impl-trait/generic_duplicate_param_use2.stderr:1]: The difference is when the checks run and which "uses" (mentions, really) are considered "defining".
From the looks of the code, "defining" vs not is an unimplemented part of the feature.
At least now `check::wfcheck` and `type_of` are each more uniform and it should be easier to finish the feature, but I'd rather leave that to someone else.

File: src/test/ui/type-alias-impl-trait/generic_nondefining_use.stderr
Comment (eddyb) [reply src/test/ui/type-alias-impl-trait/generic_nondefining_use.stderr:13]: I used the error that used to be only emitted for lifetimes, also for types and consts.

I don't like either error but I didn't want to make a third variant of my own.

File: src/librustc_typeck/collect/type_of.rs
Comment (eddyb) [inline src/librustc_typeck/collect/type_of.rs:400]: @nikomatsakis Hang on, did you see this, or not? GitHub doesn't show it on your original comment, maybe I added it afterwards?

Comment (eddyb) [reply src/librustc_typeck/collect/type_of.rs:400]: Hmm, I haven't pushed since, maybe you left the comment on a commit before I had added the comment?

File: src/librustc_typeck/collect/type_of.rs
Comment (eddyb) [reply src/librustc_typeck/collect/type_of.rs:416]: Actually, @matthewjasper, shouldn't `ReFree` turn back into `ReLateBound`? Although, I guess the actual concrete type is in terms of the `type Foo<...> = impl Trait;` generics, so it's not affected by the substitutions back at the point `Foo` is mentioned.

Comment (matthewjasper) [reply src/librustc_typeck/collect/type_of.rs:416]: I don't think that we currently ever go from a placeholder back to a bound region/type.

**Pull request merged**
